llarp/link/iwp.cpp

```diff
@@ -1,25 +1,165 @@
+                         LinkMessageHandler h, SessionEstablishedHandler est,
+                         SessionRenegotiateHandler reneg, SignBufferFunc sign,
+                         TimeoutHandler t, SessionClosedHandler closed)
+        : ILinkLayer(enckey, getrc, h, sign, est, reneg, t, closed), crypto(c)
+    {
+    }
+
+    LinkLayer::~LinkLayer()
+    {
+    }
+
+    void
+    LinkLayer::Pump()
+    {
+      ILinkLayer::Pump();
+    }
+
+    const char*
+    LinkLayer::Name() const
+    {
+      return "iwp";
+    }
+
+    bool
+    LinkLayer::KeyGen(SecretKey& k)
+    {
+      k.Zero();
+      crypto->encryption_keygen(k);
+      return !k.IsZero();
+    }
+
+    uint16_t
+    LinkLayer::Rank() const
+    {
+      return 2;
+    }
+
+    bool
+    LinkLayer::Start(Logic* l)
+    {
+      if(!ILinkLayer::Start(l))
+        return false;
+      /// TODO: change me to true when done
+      return false;
+    }
+
+    void
+    LinkLayer::RecvFrom(const Addr& from, const void* pkt, size_t sz)
+    {
+      m_OuterMsg.Clear();
+      llarp_buffer_t buf(pkt, sz);
+      if(!m_OuterMsg.Decode(&buf))
+      {
+        LogError("failed to decode outer message");
+        return;
+      }
+      NetID ourNetID;
+      switch(m_OuterMsg.command)
+      {
+        case eOCMD_ObtainFlowID:
+          if(!ShouldSendFlowID(from))
+            return;  // drop
+
+          if(m_OuterMsg.netid == ourNetID)
+            SendFlowID(from, m_OuterMsg.flow);
+          else
+            SendReject(from, m_OuterMsg.flow, "bad net id");
+      }
+    }
+
     std::unique_ptr< ILinkLayer >
-    NewServerFromRouter(AbstractRouter*)
+    NewServerFromRouter(AbstractRouter* r)
     {
-      // TODO: implement me
-      return nullptr;
+      using namespace std::placeholders;
+      return NewServer(
+          r->crypto(), r->encryption(), std::bind(&AbstractRouter::rc, r),
+          std::bind(&AbstractRouter::HandleRecvLinkMessageBuffer, r, _1, _2),
+          std::bind(&AbstractRouter::OnSessionEstablished, r, _1),
+          std::bind(&AbstractRouter::CheckRenegotiateValid, r, _1, _2),
+          std::bind(&AbstractRouter::Sign, r, _1, _2),
+          std::bind(&AbstractRouter::OnConnectTimeout, r, _1),
+          std::bind(&AbstractRouter::SessionClosed, r, _1));
     }
 
     std::unique_ptr< ILinkLayer >
-    NewServer(llarp::Crypto*, const SecretKey&, llarp::GetRCFunc,
-              llarp::LinkMessageHandler, llarp::SessionEstablishedHandler,
-              llarp::SessionRenegotiateHandler, llarp::SignBufferFunc,
-              llarp::TimeoutHandler, llarp::SessionClosedHandler)
+    NewServer(Crypto* c, const SecretKey& enckey, GetRCFunc getrc,
+              LinkMessageHandler h, SessionEstablishedHandler est,
+              SessionRenegotiateHandler reneg, SignBufferFunc sign,
+              TimeoutHandler t, SessionClosedHandler closed)
     {
-      // TODO: implement me
-      return nullptr;
+      return std::unique_ptr< ILinkLayer >(
```

Comment:
FYI `return std::make_unique<LinkLayer>(c, .....);` is a safer option in the general case (doesn't make a difference here)

Reply:
this code is not done yet.